montblanc/impl/rime/tensorflow/rime_ops/zernike_op_cpu.h

```diff
@@ -0,0 +1,152 @@
+#ifndef ZERNIKE_DDE_ZERNIKE_OP_CPU_H
+#define ZERNIKE_DDE_ZERNIKE_OP_CPU_H
+
+#include "zernike_op.h"
+#include <math.h>
+
+// Required in order for Eigen::ThreadPoolDevice to be an actual type
+#define EIGEN_USE_THREADS
+
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+
+MONTBLANC_NAMESPACE_BEGIN
+MONTBLANC_ZERNIKE_NAMESPACE_BEGIN
+
+// For simpler partial specialisation
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+// Specialise the Zernike op for CPUs
+template <typename FT, typename CT>
+class Zernike<CPUDevice, FT, CT> : public tensorflow::OpKernel
+{
+public:
+    explicit Zernike(tensorflow::OpKernelConstruction * context) :
+        tensorflow::OpKernel(context) {}
+
+    void Compute(tensorflow::OpKernelContext * context) override
+    {
+        namespace tf = tensorflow;
+
+        // Create reference to input Tensorflow tensors
+        const auto & in_coords = context->input(0);
+        const auto & in_coeffs = context->input(1);
+        const auto & in_noll_index = context->input(2);
+        const auto & in_pointing_error = context->input(3);
+        const auto & in_antenna_scaling = context->input(4);
+        
+
+        // Extract Eigen tensors
+        auto coords = in_coords.tensor<FT, 3>();
+        auto coeffs = in_coeffs.tensor<CT, 4>();
+        auto noll_index = in_noll_index.tensor<tensorflow::int32, 4>();
+        auto pointing_error = in_pointing_error.tensor<FT, 5>();
+        auto antenna_scaling = in_antenna_scaling.tensor<FT, 4>();
+
+        int nsrc = in_coords.dim_size(0);
+        int ntime = in_pointing_error.dim_size(0);
+        int na = in_coeffs.dim_size(0);
+        int nchan = in_coeffs.dim_size(1);
+        int npoly = in_coeffs.dim_size(2);
+        
+
+        // Allocate output tensors
+        // Allocate space for output tensor 'zernike_value'
+        tf::Tensor * zernike_value_ptr = nullptr;
+        tf::TensorShape zernike_value_shape = tf::TensorShape({ 
+            nsrc, 
+            ntime, 
+            na, 
+            nchan, 
+            4 });
+        OP_REQUIRES_OK(context, context->allocate_output(
+            0, zernike_value_shape, &zernike_value_ptr));
+        auto zernike_value = zernike_value_ptr->tensor<CT, 5>();
+
+        #pragma omp parallel for
+        for(int corr = 0; corr < 4 ; corr++){
+            #pragma omp parallel for
+            for(int src = 0; src < nsrc; src++){
+                FT l = coords(src, corr,0 );
+                FT m = coords(src, corr, 1);
+
```

Comment:
The l and m coordinates are not modified by pointing_errors or antenna_scaling and I think your CPU and GPU will not always agree. The test cases set them to [Identity Elements](https://en.wikipedia.org/wiki/Identity_element) which I think are very appropriate for the test cases which test that the correct expansion of the Zernike polynomials occur.

The solution is to construct another test case taking random, or semi-random inputs and which tests that the CPU and the GPU versions agree, but which does not test the correct polynomial expansion.